WordPress/src/main/java/org/wordpress/android/ui/mysite/CurrentAvatarSource.kt

```diff
@@ -14,18 +14,34 @@ import javax.inject.Singleton
 class CurrentAvatarSource
 @Inject constructor(private val accountStore: AccountStore) : SiteIndependentSource<CurrentAvatarUrl> {
     private val avatarUrl = MutableLiveData<CurrentAvatarUrl>()
+    val refresh: MutableLiveData<Boolean> = MutableLiveData<Boolean>(false)
+
     override fun buildSource(coroutineScope: CoroutineScope): LiveData<CurrentAvatarUrl> {
         val result = MediatorLiveData<CurrentAvatarUrl>()
-        result.value = CurrentAvatarUrl(accountStore.account?.avatarUrl.orEmpty())
-        result.addSource(avatarUrl) {
-            result.value = it
+        result.refreshData(false)
+        result.addSource(refresh) {
+            if (refresh.value == true) {
+                result.refreshData(true)
```

Comment:
Suggestion (💡): What if we could make this api a bit more straightforward by introducing a new `refresh` state, the `null` state for when there is no refresh involved. See below suggestion:

```
    override fun buildSource(coroutineScope: CoroutineScope): LiveData<CurrentAvatarUrl> {
        val result = MediatorLiveData<CurrentAvatarUrl>()
        result.refreshData()
        result.addSource(refresh) {
            result.refreshData(refresh.value)
        }
        return result
    }

    private fun MediatorLiveData<CurrentAvatarUrl>.refreshData(
        isRefresh: Boolean? = null
    ) {
        val url = accountStore.account?.avatarUrl.orEmpty()
        when (isRefresh) {
            null -> this@refreshData.postValue(CurrentAvatarUrl(url))
            true -> {
                refresh.postValue(false)
                if (url != avatarUrl.value?.url) {
                    this@refreshData.postValue(CurrentAvatarUrl(url))
                }
            }
            false -> Unit // Do nothing
        }
    }
```

I kind of dedicate some time to think on that since any solution that we are going to end up picking might become the go-to solution for all the other sources as well, and that is this is since we really want to avoid doing any refresh when the value of the refresh `LiveData` is `false`.

I would consider the above suggested solution for 3 reasons:
- It makes the client calls more straightforward:
  - During initial build, you can just call `refreshData()` without caring about the `refresh` parameter or its value, it is just `null`.
  - During a refresh, then you don't care about checking whether the refresh value is `true` or setting the `refresh` parameter to `true`, you just pass it through to the `refreshData()` method to do it's work.
  - The `refreshData()` makes it obvious that `null` is for the initial build, `true` is for refreshing and `false` is ignored.

Wdyt Annmarie?

Reply:
@ParaskP7 - My feelings about a Boolean having more than true/false states defeats the purpose of actually using a Boolean. This will have implications later on down to road and we'd also to manage expectations around how to treat that `null` value when checking for isRefreshing.

Reply:
> and that is this is since we really want to avoid doing any refresh when the value of the refresh LiveData is false.

Yes, that's why if it's "true" is when we refresh. I'm not comfortable introducing a null. It feels like it's less straightforward than just checking for true.

Reply:
👋 @zwarm !

Thanks for all your arguments, I totally get all your points. 💯 

It is just that I am being skeptical about building the source to require the `refresh` parameter. In my mind building a source shouldn't have to deal with refresh at all, it should just do the building. Image how the `buildSource(...)` functionality will look like if no refresh was presented or allowed. This branching of building the data wouldn't care about it. Maybe the named we are using, the `refreshData(...)` is what is misguiding us to think that this functionality is based on refresh. Maybe using the `buildData(...)` might have been a better naming, then, when `refresh` enters the game, then the `buildData()` will work for the logical branch that don't care about refresh (`null`), and for the branch that does (`true/false`).

Thus, having `null` or the `unavailable` refresh value for when that happens makes sense to me, always in my mind, I am not trying to say this is correct or the best solution. Using a state as you suggested, with `initial`, `refreshing` and `not-refreshing` might be another way to tackle that. It does feel a bit more involved that a fixed 3 value only nullable boolean solution, and it might suggest that more states might be allowed, but it might better way of dealing with this problem. 🤷 

I thinking about it out loud because I am not that comfortable with the below solution that every client/source will need to repeat, always needing an extra `refresh` parameter to work:

```
        result.refreshData(false)
        result.addSource(refresh) {
            if (refresh.value == true) {
                result.refreshData(true)
            }
        }
```

Also, for this extra `if -> true` logic, which we should be careful as we will then need to match the `true` for the checking and for passing the value down to `refreshData(...)`.

PS.1: If you are comfortable with this solution, by all means let's use that. 👍 
PS.2: Maybe we can involved @ashiagr here as well so that we can agree on it and move forward. 🤔

Reply:
> Image how the buildSource(...) functionality will look like if no refresh was presented or allowed

Yes, it would look exactly as it did before we introduced `refresh`. But now we have to introduce refresh and it's going to be in that same method. 

As a compromise, I am changing to the following syntax. This eliminates the `if (refresh.value == true) ` check and makes the code tidier.

override fun buildSource(coroutineScope: CoroutineScope): LiveData<CurrentAvatarUrl> {
        val result = MediatorLiveData<CurrentAvatarUrl>()
        result.refreshData()
        result.addSource(refresh) {
            result.refreshData(refresh.value)
        }
        return result
    }

    private fun MediatorLiveData<CurrentAvatarUrl>.refreshData(
        isRefresh: Boolean? = null
    ) {
        val url = accountStore.account?.avatarUrl.orEmpty()
        when (isRefresh) {
            true -> {
                refresh.postValue(false)
                if (url != avatarUrl.value?.url) {
                    this@refreshData.postValue(CurrentAvatarUrl(url))
                }
            }
            false -> this@refreshData.postValue(CurrentAvatarUrl(url))
        }
    }

Reply:
@ParaskP7 - As discussed, using a `Boolean?` in an isolated method (as in this instance) works for me. Thanks for the suggestion.

Reply:
Updated in 11b0c3f 🙇  @ParaskP7 - if you wanted to take another look.

Reply:
👋 @zwarm !

Thanks so much for the discussion and for the change, everything LGTM, still a 🟢 from my side!

🏅